python/cudaq/qis/qis.py

```diff
@@ -0,0 +1,271 @@
+            qubitIds.append(a.id())
+        elif isinstance(a, qvector) or isinstance(a, qview):
+            [qubitIds.append(q.id()) for q in a]
+        else:
+            raise Exception(
+                "invalid argument type passed to {}.__call__".format(opName))
+    return qubitIds
+
+
+class quantum_operation(object):
+    """
+    A quantum_operation provides a base class interface for invoking 
+    a specific quantum gate, as well as controlled and adjoint versions 
+    of the gate.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def get_name():
+        """
+        Return the name of this operation.
+        """
+        pass
+
+    @classmethod
+    def get_num_parameters(cls):
+        """
+        Return the number of rotational parameters this operation requires.
+        """
+        return 0
+
+    @classmethod
+    def __call__(cls, *args):
+        """
+        Invoke the quantum operation. The arguments can contain float parameters (of the
+        correct number according to `get_num_parameters`) and quantum types (`qubit`, `qvector`, `qview`).
+        """
+        opName = cls.get_name()
+        parameters = list(args)[:cls.get_num_parameters()]
+        quantumArguments = list(args)[cls.get_num_parameters():]
+        qubitIds = [q for q in processQubitIds(opName, *quantumArguments)]
+
+        [
+            cudaq_runtime.applyQuantumOperation(opName, parameters, [], [q],
+                                                False, SpinOperator())
+            for q in qubitIds
+        ]
+
+    @classmethod
+    def ctrl(cls, *args):
+        """
+        Invoke the general controlled version of the quantum operation. 
+        The arguments can contain float parameters (of the correct number according
+        to `get_num_parameters`) and quantum types (`qubit`, `qvector`, `qview`).
+        """
+        opName = cls.get_name()
+        parameters = list(args)[:cls.get_num_parameters()]
+        quantumArguments = list(args)[cls.get_num_parameters():]
+        qubitIds = processQubitIds(opName, *quantumArguments)
+        controls = qubitIds[:len(qubitIds) - 1]
+        targets = [qubitIds[-1]]
+
+        for q in quantumArguments:
+            if isinstance(q, qubit) and q.is_negated():
+                x()(q)
+
+        cudaq_runtime.applyQuantumOperation(opName, parameters, controls,
+                                            targets, False, SpinOperator())
+        for q in quantumArguments:
+            if isinstance(q, qubit) and q.is_negated():
+                x()(q)
+                q.reset_negation()
+
+    @classmethod
+    def adj(cls, *args):
+        """
+        Invoke the general adjoint version of the quantum operation. 
+        The arguments can contain float parameters (of the correct number according
+        to `get_num_parameters`) and quantum types (`qubit`, `qvector`, `qview`).
+        """
+        opName = cls.get_name()
+        parameters = list(args)[:cls.get_num_parameters()]
```

Comment:
Ah, that's what you use them for. 
Can't we look at the type of the parameter instead? 

More broadly, when are these implementations (ctrl and adj) executed? I.e. isn't there direct support for them in the ast bridge? If we had support before this change, what additional support does this enable or replace (I don't see anything removed)?

Is this implementation here to support the existing `c<op>` (e.g. cx) function name? If that's the purpose, why not simply translate it into <op>.ctrl here and let the compiler handle dealing with it?

Reply:
I can answer these questions: 

None of these functions will ever be executed - they were from the eager mode implementation. They are potentially useful here for the Pylance problems since they define all our operations + docstrings succinctly. 

"Can't we look at the type of the parameter" - you could do that, but it would be messier than the current implementation, since you'll have to insert a number of `if isinstance(...)` calls, here you just have the information on the operation class and ask it for that information.

Reply:
I think that's a good idea